# Regression Model Using Keras
Trying Keras with different parameter to analyze the result

In this Notebook will be divide to 5 model with different parameter

In [ ]:
import pandas as pd
df = pd.read_csv('https://cocl.us/concrete_data')

In [ ]:
df.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

Explore the data

In [ ]:
df.dtypes

Cement                float64
Blast Furnace Slag    float64
Fly Ash               float64
Water                 float64
Superplasticizer      float64
Coarse Aggregate      float64
Fine Aggregate        float64
Age                     int64
Strength              float64
dtype: object

In [ ]:
df.isnull().any()

Cement                False
Blast Furnace Slag    False
Fly Ash               False
Water                 False
Superplasticizer      False
Coarse Aggregate      False
Fine Aggregate        False
Age                   False
Strength              False
dtype: bool

In [ ]:
df.describe()

Cement  Blast Furnace Slag  ...          Age     Strength
count  1030.000000         1030.000000  ...  1030.000000  1030.000000
mean    281.167864           73.895825  ...    45.662136    35.817961
std     104.506364           86.279342  ...    63.169912    16.705742
min     102.000000            0.000000  ...     1.000000     2.330000
25%     192.375000            0.000000  ...     7.000000    23.710000
50%     272.900000           22.000000  ...    28.000000    34.445000
75%     350.000000          142.950000  ...    56.000000    46.135000
max     540.000000          359.400000  ...   365.000000    82.600000

[8 rows x 9 columns]

## Divide between Features and Target

In [ ]:
df_columns = df.columns

predictors = df[df_columns[df_columns != 'Strength']] # all columns except Strength
target = df['Strength'] # Strength column

In [ ]:
n_cols = predictors.shape[1]

## **Keras Model Task A**

In [ ]:
import keras
from keras.layers import Dense
from keras.models import Sequential

In [ ]:
def regression_model():
  model = Sequential()
  model.add(Dense(10,activation = 'relu', input_shape = (n_cols,)))
  model.add(Dense(1))
  #
  model.compile(optimizer = 'adam', loss = 'mean_squared_error')
  return model

## Keras Model Task A Number 1
Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size = 0.3, random_state = 1)

## Keras Model Task A Number 2
Train the model on the training data using 50 epochs.



In [ ]:
model = regression_model()
model.fit(X_train,y_train,epochs = 50)

Epoch 1/50
23/23 [==============================] - 15s 1ms/step - loss: 2467.4980
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 1325.7583
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 1079.9244
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 904.1592
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 861.9161
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 853.4550
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 813.8265
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 651.1402
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 661.6015
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 579.4225
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 548.5602
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 506.4496
Epoch 13/50
23/23 [==

## Keras Model Task A Number 3
Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [ ]:
from sklearn.metrics import mean_squared_error
y_hat = model.predict(X_test)
print(mean_squared_error(y_test, y_hat))


152.4505585833854


The Mean Square Error is 152.45


## Keras Model Task A Number 4
Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [ ]:
MSE_List = []
for x in range(50):
  model = regression_model()
  model.fit(X_train,y_train,epochs = 50)
  y_hat = model.predict(X_test)
  loss = mean_squared_error(y_test, y_hat)
  MSE_List.append((x,loss))

print(MSE_List)




Streaming output truncated to the last 5000 lines.
23/23 [==============================] - 1s 1ms/step - loss: 15729.7363
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 5627.7490
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 5180.4758
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 4531.0207
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 4077.0845
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 3847.5551
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 3103.3248
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 2801.4434
Epoch 9/50
23/23 [==============================] - 0s 1ms/step - loss: 2584.5798
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 2143.5151
Epoch 11/50
23/23 [==============================] - 0s 1ms/step - loss: 1781.2711
Epoch 12/50
23/23 [==============================] - 0s

## Keras Model Task A Number 5
Report the mean and the standard deviation of the mean squared errors.

In [ ]:
import numpy as np
print("The Mean of MSE is {}".format(np.mean(MSE_List)))
print("The Std of MSE is {}".format(np.std(MSE_List)))

The Mean of MSE is 212.2915079987289
The Std of MSE is 522.6325751282347


## **Keras Model Task B**
In this model will normalize the data using help of StandardScaler from sklearn.
Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation and its the same as StandardScaler from sklearn

In [ ]:
import sklearn
from sklearn.preprocessing import StandardScaler

In [ ]:
SS = StandardScaler()
X_train_norm = SS.fit_transform(X_train)
X_test_norm = SS.transform(X_test)

In [ ]:
MSE_List = []
for x in range(50):
  model = regression_model()
  model.fit(X_train_norm,y_train,epochs = 50)
  y_hat = model.predict(X_test_norm)
  loss = mean_squared_error(y_test, y_hat)
  MSE_List.append((loss))

Streaming output truncated to the last 5000 lines.
Epoch 1/50
23/23 [==============================] - 1s 1ms/step - loss: 1534.4052
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 1581.5618
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 1542.3483
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 1502.4088
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 1459.4615
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 1444.0159
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 1402.2646
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 1469.9873
Epoch 9/50
23/23 [==============================] - 0s 1ms/step - loss: 1318.0356
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 1365.1001
Epoch 11/50
23/23 [==============================] - 0s 1ms/step - loss: 1440.8524
Epoch 12/50
23/23 [==========================

In [ ]:
print("The Mean of MSE is {}".format(np.mean(MSE_List)))
print("The Std of MSE is {}".format(np.std(MSE_List)))

The Mean of MSE is 382.3070536773194
The Std of MSE is 116.50852908061294


## Task B Answer
Task A
The Mean of MSE is 212.2
The Std of MSE is 522.6

Task B
The Mean of MSE is 382.3
The Std of MSE is 116.5


While Task A has smaller Mean of MSE, Task B has small std that mean its more precision than before

## **Keras Model Task C**
Repeat Part B but use 100 epochs this time for training.

In [24]:
SS = StandardScaler()
X_train_norm = SS.fit_transform(X_train)
X_test_norm = SS.transform(X_test)

In [ ]:
MSE_List = []
for x in range(50):
  model = regression_model()
  model.fit(X_train_norm,y_train,epochs = 100)
  y_hat = model.predict(X_test_norm)
  loss = mean_squared_error(y_test, y_hat)
  MSE_List.append((loss))

Streaming output truncated to the last 5000 lines.
Epoch 1/100
23/23 [==============================] - 1s 2ms/step - loss: 1577.9763
Epoch 2/100
23/23 [==============================] - 0s 2ms/step - loss: 1523.2510
Epoch 3/100
23/23 [==============================] - 0s 2ms/step - loss: 1456.0463
Epoch 4/100
23/23 [==============================] - 0s 2ms/step - loss: 1574.1863
Epoch 5/100
23/23 [==============================] - 0s 2ms/step - loss: 1457.9142
Epoch 6/100
23/23 [==============================] - 0s 2ms/step - loss: 1497.2201
Epoch 7/100
23/23 [==============================] - 0s 2ms/step - loss: 1441.6448
Epoch 8/100
23/23 [==============================] - 0s 2ms/step - loss: 1439.6425
Epoch 9/100
23/23 [==============================] - 0s 2ms/step - loss: 1509.7458
Epoch 10/100
23/23 [==============================] - 0s 2ms/step - loss: 1381.3370
Epoch 11/100
23/23 [==============================] - 0s 2ms/step - loss: 1421.4715
Epoch 12/100
23/23 [==============

In [ ]:
print("The Mean of MSE is {}".format(np.mean(MSE_List)))
print("The Std of MSE is {}".format(np.std(MSE_List)))

The Mean of MSE is 190.30073964512823
The Std of MSE is 12.40566689819291


## Task C Answer
Task A
The Mean of MSE is 212.2
The Std of MSE is 522.6

Task B
The Mean of MSE is 382.3
The Std of MSE is 116.5

Task C
The Mean of MSE is 190.3
The Std of MSE is 12.4


Task C has smaller Mean and Std of MSE if compared to Task B.
The result is better

## **Keras Model Task D**
Repeat Part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

In [25]:
SS = StandardScaler()
X_train_norm = SS.fit_transform(X_train)
X_test_norm = SS.transform(X_test)

In [ ]:
def regression_model():
  model = Sequential()
  model.add(Dense(10,activation = 'relu', input_shape = (n_cols,)))
  model.add(Dense(10,activation = 'relu', input_shape = (n_cols,)))
  model.add(Dense(10,activation = 'relu', input_shape = (n_cols,)))
  model.add(Dense(1))
  #
  model.compile(optimizer = 'adam', loss = 'mean_squared_error')
  return model

MSE_List = []
for x in range(50):
  model = regression_model()
  model.fit(X_train_norm,y_train,epochs = 50)
  y_hat = model.predict(X_test_norm)
  loss = mean_squared_error(y_test, y_hat)
  MSE_List.append((x,loss))

Streaming output truncated to the last 5000 lines.
Epoch 1/50
23/23 [==============================] - 1s 2ms/step - loss: 1575.5051
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 1544.0450
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 1433.3053
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 1455.4121
Epoch 5/50
23/23 [==============================] - 0s 2ms/step - loss: 1431.7134
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 1385.4028
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 1282.2545
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 1228.7215
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 1028.3414
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 760.6619
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 603.3388
Epoch 12/50
23/23 [============================

In [ ]:
print("The Mean of MSE is {}".format(np.mean(MSE_List)))
print("The Std of MSE is {}".format(np.std(MSE_List)))

The Mean of MSE is 86.22699586284406
The Std of MSE is 63.24578431151329


## Task D Answer
Task A
The Mean of MSE is 212.2
The Std of MSE is 522.6

Task B
The Mean of MSE is 382.3
The Std of MSE is 116.5

Task C
The Mean of MSE is 190.3
The Std of MSE is 12.4

Task D
The Mean of MSE is 86.2
The Std of MSE is 63.2


Increasing hidden layer give better result but theres cons that the time to complete the model is way longer than before